File: 2º/PDOO/ProyectoRUBY/lib/sorpresa.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Sorpresa
    def initialize(texto="sorpresa")
      @texto=texto
    end
    def jugadorCorrecto(actual,todos)
      if(actual<=todos.size)then
        return true
      else
        return false
      end
    end
    def setValor(nuevo)
        @valor=nuevo
    end
    attr_reader :tipoSorpresa
    attr_reader :texto
    attr_reader :valor
    attr_reader :tabla
    attr_reader :mazito
    def salirDelMazo
      if(getTipoSorpresa()==TipoSorpresa.IRCARCEL)then
        @mazito.inhabilitarCartaEspecial(this)
      end
    end
    def to_s
      return @texto
    end
    def informe(actual, todos)
      if(jugadorCorrecto(actual, todos))then
        evento="Al jugador "+ todos[actual].getNombre()
        evento=evento+" se le aplicara la sorpresa #{texto}"       
        Diario.instance.ocurre_evento(evento)
      end
    end
    private :texto, :valor, :tabla, :tipoSorpresa, :mazito
  end
end

File: 2º/PDOO/ProyectoJAVA/src/civitas/Sorpresa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author nikitastetskiy
 */
public abstract class Sorpresa {
    
    private String texto;
    
    abstract void aplicarAJugador(int actual, ArrayList<Jugador> todos);
    
    void informe(int actual, ArrayList<Jugador> todos){
        String e="El jugador "+todos.get(actual).getNombre()+" utiliza sorpresa: "+toString();
        Diario.getInstance().ocurreEvento(e);
    }
    
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
        return actual<todos.size() && actual>=0;
    }
   
    Sorpresa(String texto){
        this.texto = texto;
    }
    
    @Override
    public String toString() {
        return "<<" + texto + ">>";
    }
}